Add experimental procedural generator and Scenario visualization as graph

galigula2 opened the pull request:
Initial release of the mentioned features. The graph visualization works pretty nicely but of course could be improved further. The procedural generator is very much in experimental phase but can already be used to create something to start expanding on in the main editor. The generated scenario should be "playable" in a sense that it can be clicked through but most likely is not balanced at all.

galigula2 committed:
Add initial procedural boilerplate

galigula2 committed:
Work on Scenario visualization as graph

galigula2 committed:
Work on Scenario visualization

galigula2 committed:
Improve visualization

- still missing some end triggers in Weathertop

galigula2 committed:
Update visualization

galigula2 committed:
Visualization tooltip improvements

galigula2 committed:
Implement Objective and StoryPoint creation

galigula2 committed:
Small tweaks and exporting visualization as image

galigula2 committed:
Initial StoryArchetype  drafting

galigula2 committed:
Flesh out story generation a bit

- Also fix some basic incosistencies in the Objective graph

galigula2 committed:
Improve story generation

galigula2 committed:
Work on tile info and generation

galigula2 committed:
Further work for story elements and tiles

galigula2 committed:
Work on interactions

galigula2 committed:
Work on story mechanics

- Still something fishy going on with next objective not triggering

galigula2 committed:
Update all location tile infoes and small other tweaks

galigula2 committed:
Simple story generation not technically works!

- It can generate a scenario that can be clicked through in the app.
- Still lots of work to be done before the scenario is in any way playable.

galigula2 committed:
Split up StoryGenerator and add StoryFragment loading from file

- NOTE: Generated Scenario looks a bit odd and does not play out properly. Need to bisect!

galigula2 committed:
Implement skeleton for Stat Test interactions

galigula2 committed:
Start implementing Threat interaction

- TODO: not appearing in the application

galigula2 committed:
Improve token and context handling

galigula2 committed:
Start filling in story details

galigula2 committed:
Generate Objective texts

galigula2 committed:
Implement rest of the interaction fillers

galigula2 committed:
Implemented very simple Threat track generation

galigula2 committed:
Clean up code and TODO's (WIP)

galigula2 committed:
Clean up TODO's and small adjustments

galigula2 committed:
Make generator screen compile

- TODO: Still need to hook up parameters

galigula2 committed:
Basic parameters ready in Procedural Scenario Generator

galigula2 committed:
Allow procedural generator also for Campaigns

galigula2 committed:
Use visualization also in main editor

galigula2 committed:
Implement visualization clicks in main screen

galigula2 committed:
Update visualization on changes and small tweaks

galigula2 committed:
Implement advanced parameter and add experimental text

galigula2 committed:
Show generator logs in generator screen

galigula2 committed:
Add info tooltips to procedural generator screen

galigula2 committed:
Clean up procedural code structure

galigula2 committed:
Change StoryTemplate structure to have SupportedArchetypes

- This makes it easier to write texts in templates so that
  they match the expected archetype but still leave some
  room for re-use in parts of the template
- Also added StoryElementChecker that shows in the
  UI if there are problems (still wip)

galigula2 committed:
Add TheQuest parts for the story template

galigula2 committed:
Remove VoyageAndReturn archetype for now, add tooltips to parameter screen

galigula2 committed:
Adjust StatTest interactions

galigula2 committed:
Clean up TODO's

galigula2 committed:
Merge remote-tracking branch 'origin/master' into procedural

TrnsltLife commented on MainWindow.xaml.cs:
```diff
@@ -775,9 +881,33 @@ private void ChapterPropsButton_Click( object sender, RoutedEventArgs e )
+            OpenChapterEditor(((Button)e.Source).DataContext as Chapter);
+        }
+
+        private void OpenTokenEditor(Token t)
```
At some point these should get wired up to work for SquareTile too.

galigula2 replied:
True, I can test this with some of the test scenarios so I'll fix it right away

TrnsltLife commented on MainWindow.xaml:
```diff
@@ -4,9 +4,11 @@
         mc:Ignorable="d"
 				xmlns:uc="clr-namespace:JiME.UserControls"
-        Title="{Binding titleChangedToken, Converter={StaticResource titleConverter}}" Height="500" Width="770" MinHeight="500" MinWidth="700" WindowStartupLocation="CenterScreen" Background="#FF2D2D30" Closing="Window_Closing" ResizeMode="CanResizeWithGrip" SizeToContent="Width">
+        Title="{Binding titleChangedToken, Converter={StaticResource titleConverter}}" Height="500" Width="770" MinHeight="500" MinWidth="700" WindowStartupLocation="CenterScreen" Background="#FF2D2D30" Closing="Window_Closing" ResizeMode="CanResizeWithGrip" SizeToContent="WidthAndHeight">
```
It's good that it is filling the space, but it's filling it a little too much. It's putting the window menu bar above the top of my screen, and the window status bar below the bottom of my screen. On dual monitors, it does the same thing, so stretches to fill both monitors plus a little more. If you could fix this, that would be great.

galigula2 replied:
Agreed, I'm considering just maximizing the main editor window when it opens unless you have some objections?

TrnsltLife replied:
Yeah, let's give that a try.

TrnsltLife commented on Models/Visualization/Graph.cs:
```diff
@@ -0,0 +1,460 @@
+                }
+                else
+                {
+                    throw new NotImplementedException("Graph preparation not implemented for interaction type: " + x.GetType().Name);
```
Let's change this to a Debug.Log statement or something and not throw an Exception, because for files that already have an event like that, it crashes the program and doesn't allow them to continue. e.g. I had a file with a MultiEventInteraction in it. Or maybe this Exception just needs to be handled somewhere downstream.

galigula2 replied:
This was more for testing, I'll remove it completely and group all the interactions that don't have any extra connections in the graph (like Text, Darkness etc) to that else-block.

TrnsltLife commented:
The graph editor and the generator are very cool. Impressive amount of work. I tested things out a bit. I realize the generator and the graph are works in progress, so no complaints there, generally. Just a couple of blockers that I found regarding the window size initially, and unimplemented event types crashing the program. Once those are addressed, I'll be glad to merge it.

galigula2 commented:
> The graph editor and the generator are very cool. Impressive amount of work. I tested things out a bit. I realize the generator and the graph are works in progress, so no complaints there, generally. Just a couple of blockers that I found regarding the window size initially, and unimplemented event types crashing the program. Once those are addressed, I'll be glad to merge it.

Good points. I try to find time during the weekend to clean these up. I also can take a look at the unimplemented types. Do you have a jime-file that would bering these up? I'm not sure how some of the interactions should work.

TrnsltLife commented:
You can grab some .jime files that have different features here: https://github.com/TrnsltLife/YourJourneyScenarios/releases

galigula2 committed:
Map out needed changes

galigula2 commented:
Nice set of test scenarios! I'll do the fixes and let you know when it can be reviewed again,

galigula2 committed:
Fix based on comments in PR

- Maximuze the main editor window when opening
- Clean up unnecessary HexTile limitations
- Implement missing interactions in Graph

galigula2 commented:
OK, I've done the changes we discussed earlier in the comments. Could you check the visualization that it makes sense to you in the different scenarios? I may have missed some special meaning in some of the interactions. 

I also finally got the actual game and played couple of first chapters in Bones of Arnor. This gives me a better idea how the scenarios work and makes it easier to balance things out somewhat. 

Let's see after this is released if we can get some other people interested in trying things out as well.

galigula2 committed:
Get rid of StoryPointInteraction

- Was just for generator mid-phase visualization purposes and can be replaced with MultiEventInteraction

galigula2 commented:
In relation to the Patreon message you posted on the BGG message chain. Have you had any thought on how that relates to contributed content (such as the procedural generator and visualization in this PR)? I'm mostly interested in helping the community by creating something useful but of course if there is money involved and things gather speed then it's good to be on the same page on this kinds of things.

This is somewhat relevant also because I'm planning to use our company hackathon event to try out StoryTemplate generation using ChatGPT and there might be some things that depend on whether the output is for non-profit or something else. Most likely there is no problem with any of this but it doesn't hurt to have the discussion anyway. :)

TrnsltLife commented:
In relation to Patreon, I don't really know. I've worked on this project for about a year now, and kind of got burned out for a while, but on my Jira board I have about 30 different feature ideas, bug fixes, art ideas, and other things I think would be cool to do. Plus I have ideas on scenarios I want to craft by hand and so on.

For the time being at least it seems like we might have somewhat separate domains, although of course they have to be linked together to some extent to work, the same underlying DOM so to speak.

I don't know if anyone is even going to want to sign up for "my Patreon" or if it is a dumb idea or what. But I thought I'd give it a try and see what happened. Like if people actually sign up and support me, that would certainly give me motivation to keep going. And if not, maybe I'll realign my priorities or something. If nothing else, I've now set up my Hello World Patreon site and DIscord server, so I've learned something new.

You could always create your own Patreon and provide support and perks for the procedural generation side of things. Have people support you for creating new story prototypes, story templates, etc.? Or for the scenarios you can create using the procedural generator? Tutorials on how to use it? Or one-on-one time with you for help using it?

Another idea could be if you wanted to split out the procedural generation as a separate app that could then launch JiMEditor with the finished file in case the user wanted to do additional editing. Not that you have to do that, but if you wanted the procedural generation code kept separate for some reason.

I don't plan to claim credit for features you've added. Or put promises on Patreon that I know are things you're planning to work on, for example. I'm just asking for support for things I've done, or will do, and see where that leads.

For example, whatever code you end up wanting to put in this repo, for the visualizer or the procedural generator, when I create a new release, you could definitely post your own release notes and highlights on the things you've added, updated, with screenshots, etc.

What are your thoughts on this?

TrnsltLife commented:
I think having the editor maximize the window works great.

I'm glad I can click 999A and have it open the Battle Map tile now.

Here's something to look at:
![Visualizer New Area Question](https://user-images.githubusercontent.com/280842/232425676-f32eedbb-3872-4a7b-bc13-c0daf4c6164f.png)

In the "Search for the Citadel" scenario, there's a Search Token on tile 300A. When you click that, it starts the New Area event. The New Area event, when it is done, calls the New Area trigger (same name, different item). The New Area trigger then causes the New Area tile block to be placed on the map.

I'm a little confused about the graph though. It seems to show both the Event and the Tile feeding into the Trigger, but I'm not sure that's right:
```
(New Area event)------------------------\
                                         }---->(New Area trigger)
(300A)---->(Search)---->(triggers)-----/
```

I think there may be other places that we'll find have issues, since some of the Event types have multiple optional triggers feeding into them, or multiple events or triggers coming out of them. But this is the one I noticed so far.

I also want to say it doesn't have to be perfect before we merge, since sometimes things are good enough to be useful and also start getting feedback.

galigula2 commented:
> In the "Search for the Citadel" scenario, there's a Search Token on tile 300A. When you click that, it starts the New Area event. The New Area event, when it is done, calls the New Area trigger (same name, different item). The New Area trigger then causes the New Area tile block to be placed on the map.
> I'm a little confused about the graph though. It seems to show both the Event and the Tile feeding into the Trigger, but I'm not sure that's right:

I saw this while testing but I interpreted it as a obsolete leftover. The reason why it is shown like that in the graph is that 
- There is no "Triggered By" so nothing actually triggers this directly. This is natural since this is a token interaction but it seems that the token is not placed under any tile? Or perhaps I'm missing some route where the token gets placed on to a tile randomly and that should be added to the visualization?
- The reason we get the arrow to "New area trigger" is because that is set as "Set Trigger After Event is Resolved". I interpreted that this will never happen because the token for "New area event" is never shown to the user but that might be a wrong assumption as said above. 

> I also want to say it doesn't have to be perfect before we merge, since sometimes things are good enough to be useful and also start getting feedback.

Agreed

galigula2 commented:
> You could always create your own Patreon and provide support and perks for the procedural generation side of things. Have people support you for creating new story prototypes, story templates, etc.? Or for the scenarios you can create using the procedural generator? Tutorials on how to use it? Or one-on-one time with you for help using it?

Interesting idea. Too busy with family, day job etc. for now unless interest for this explodes out of control but interested to also see how your Patreon takes off.
 
> Another idea could be if you wanted to split out the procedural generation as a separate app that could then launch JiMEditor with the finished file in case the user wanted to do additional editing. Not that you have to do that, but if you wanted the procedural generation code kept separate for some reason.

I was planning that the generator could be used directly from the Your Journey application if it ever get's to a state where the generated scenarios are playable without modification. This would require that we'd split the core model and procedural code in another Project. Could be something like JiMECore (.NET Standard 2, contains current Common, Models and Procedural classes) that is referenced by the JiMEeditor project and could also be referenced inside the Unity project. There might be some .NET version problems with this and we'd perhaps need to think about removing some of the editor logic from the Model-classes so they would have fewer dependencies. 

If this Project-level split is made then the generator code could also be easily split out but don't see the reason for that at least not yet. I'm in favor of keeping things simple as far as possible since this is for community's benefit. 

> I don't plan to claim credit for features you've added. Or put promises on Patreon that I know are things you're planning to work on, for example. I'm just asking for support for things I've done, or will do, and see where that leads.
> 
> For example, whatever code you end up wanting to put in this repo, for the visualizer or the procedural generator, when I create a new release, you could definitely post your own release notes and highlights on the things you've added, updated, with screenshots, etc.

I'd say one set of release notes is clearer for the users if we can coordinate them properly. Perhaps add something like this to the repository? https://keepachangelog.com/ Basically just a file with [Unreleased] section where we could both add details in the commits where the changes are implemented and then [Unreleased] is changed to [Release x.y.z] and new [Unreleased] section is created.

> What are your thoughts on this?

Let's see where this develops. I don't mind contributing to a free-to-use community project and if you (or me for that matter) want to have Patreon for extra support and other content I don't see a problem with that as long as the product itself remains free to use as you mentioned also in the BGG post.

galigula2 commented:
I'll do couple of checks still for the visualization.
- When clicking a token it should pre-select the correct token in the tile editor
- When clicking a chapter it should open the Chapter Properties-dialog instead of the Tile Editor for the chapter
- I didn't yet find a reason how the "New Area"-TextEvent token would be shown? Or is it shown at all? The visualization is also a good way to detect this kind of details that don't hook up to the story itself (other example would be the "River"-TextEvent in the same scenario.

galigula2 committed:
Visualization tweaks

- Open chapter properties when chapter node is clicked
- Pre-select the clicked token if token node is clicked (clicking tile node still pre-selects the first token if any)
- Make token preselection actually show the red highlight color when the token editor is opened

galigula2 commented:
> I'll do couple of checks still for the visualization.
> 
> * When clicking a token it should pre-select the correct token in the tile editor
> * When clicking a chapter it should open the Chapter Properties-dialog instead of the Tile Editor for the chapter
> * I didn't yet find a reason how the "New Area"-TextEvent token would be shown? Or is it shown at all? The visualization is also a good way to detect this kind of details that don't hook up to the story itself (other example would be the "River"-TextEvent in the same scenario.

OK, the token clicking changes implemented during lunch break. :)

As long as the "New Area"-TextEvent thing gets sorted out I think this should be ready to merge.

galigula2 committed:
Fix one place where getTriggerOrEventVertex caused problems

- Need to check all places where this is called and check if it should be limited to just triggers or events to avoid problem when dataname is identical

galigula2 commented:
> As long as the "New Area"-TextEvent thing gets sorted out I think this should be ready to merge.

Actually fixed this too, I have logic in Graph.cs that looks for vertex to connect to and it  first checks triggers and then events/interactions. In the Citadel scenario they had the same name so it selected the trigger instead of event. The UI does not allow triggers in that selection so we can safely always select the event in this case. 

Need at some point to go through all places where getTriggerOrEventVertex() is called to determine whether we can decide which one is the valid choice.

TrnsltLife commented:
Re: New Area
I was getting ready to write that I had discovered something about this, but it looks like you already figured it out. I just noticed the same thing, that if I change the Event name to "New Area Event", the graph displayed correctly; same thing if I change the Trigger name to "New Area Trigger". It's cool how the graph redraws itself so easily.

Re: Changelog
Sure, go ahead and add a CHANGELOG.md at the root level of the project and we can start using that to help coordinate.

galigula2 committed:
Explicitly select trigger or event vertex in graph.cs

galigula2 committed:
Make edge labels more distinguishable from vertex labels

galigula2 committed:
Add CHANGELOG.md

galigula2 commented:
> Changelog

I added the CHANGELOG.md with my notes for the unreleased version. I didn't yet add all the text about semantic versioning or the links the same way those were used e.g. here https://github.com/olivierlacan/keep-a-changelog/blob/main/CHANGELOG.md 

> New area

I explicitly set whether triggers or events should be searched for everywhere in graph.cs. It might still show something wrong but not at least it always searches from one set of items.

I also tweaked style of the edge labels a bit to make them more distinguishable from the vertex labels.
I think now everything we've been discussing has been fixed.

galigula2 committed:
Update procedural generator next milestones

TrnsltLife commented:
I'm OK with not bringing in semantic versioning anyways for now so that's no problem.

The sideways edge labels look more congruent now, i.e. they look like they're labeling the arrow instead of interrupting it.

galigula2 commented:
I'll continue to use the same "procedural" branch for further development.
